// @sinonjs/fake-timers/dist/bcrypt.js
/*
 Copyright (c) 2012 Nevins Bartolomeo <nevins.bartolomeo@gmail.com>
 Copyright (c) 2012 Shane Girish <shaneGirish@gmail.com>
 Copyright (c) 2014 Daniel Wirtz <dcode@dcode.io>

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:
 1. Redistributions of source code must retain the above copyright
 notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
 notice, this list of conditions and the following disclaimer in the
 documentation and/or other materials provided with the distribution.
 3. The name of the author may not be used to endorse or promote products
 derived from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/**
 * @license bcrypt.js (c) 2013 Daniel Wirtz <dcode@dcode.io>
 * Released under the Apache License, Version 2.0
 * see: https://github.com/dcodeIO/bcrypt.js for details
 */
(function(global, factory) {

    /* AMD */ if (typeof define === 'function' && define["amd"])
        define([], factory);
    /* CommonJS */ else if (typeof require === 'function' && typeof module === "object" && module && module["exports"])
        module["exports"] = factory();
    /* Global */ else
        (global["dcodeIO"] = global["dcodeIO"] || {})["bcrypt"] = factory();

}(this, function() {
    "use strict";

    /**
     * bcrypt namespace.
     * @type {Object.<string,*>}
     */
    var bcrypt = {};

    /**
     * The random implementation to use as a fallback.
     * @type {?function(number):!Array.<number>}
     * @inner
     */
    var randomFallback = null;

    /**
     * Generates cryptographically secure random bytes.
     * @function
     * @param {number} len Bytes length
     * @returns {!Array.<number>} Random bytes
     * @throws {Error} If no random implementation is available
     * @inner
     */
    function random(len) {
        /* node */ if (typeof module !== 'undefined' && module && module['exports'])
            try {
                return require("crypto")['randomBytes'](len);
            } catch (e) {}
        /* WCA */ try {
            var a; (self['crypto']||self['msCrypto'])['getRandomValues'](a = new Uint32Array(len));
            return Array.prototype.slice.call(a);
        } catch (e) {}
        /* fallback */ if (!randomFallback)
            throw Error("Neither WebCryptoAPI nor a crypto module is available. Use bcrypt.setRandomFallback to set an alternative");
        return randomFallback(len);
    }

    // Test if any secure randomness source is available
    var randomAvailable = false;
    try {
        random(1);
        randomAvailable = true;
    } catch (e) {}

    // Default fallback, if any
    randomFallback = null;
    /**
     * Sets the pseudo random number generator to use as a fallback if neither node's `crypto` module nor the Web Crypto
     *  API is available. Please note: It is highly important that the PRNG used is cryptographically secure and that it
     *  is seeded properly!
     * @param {?function(number):!Array.<number>} random Function taking the number of bytes to generate as its
     *  sole argument, returning the corresponding array of cryptographically secure random byte values.
     * @see http://nodejs.org/api/crypto.html
     * @see http://www.w3.org/TR/WebCryptoAPI/
     */
    bcrypt.setRandomFallback = function(random) {
        randomFallback = random;
    };

    /**
     * Synchronously generates a salt.
     * @param {number=} rounds Number of rounds to use, defaults to 10 if omitted
     * @param {number=} seed_length Not supported.
     * @returns {string} Resulting salt
     * @throws {Error} If a random fallback is required but not set
     * @expose
     */
    bcrypt.genSaltSync = function(rounds, seed_length) {
        rounds = rounds || GENSALT_DEFAULT_LOG2_ROUNDS;
        if (typeof rounds !== 'number')
            throw Error("Illegal arguments: "+(typeof rounds)+", "+(typeof seed_length));
        if (rounds < 4)
            rounds = 4;
        else if (rounds > 31)
            rounds = 31;
        var salt = [];
        salt.push("$2a$");
        if (rounds < 10)
            salt.push("0");
        salt.push(rounds.toString());
        salt.push('$');
        salt.push(base64_encode(random(BCRYPT_SALT_LEN), BCRYPT_SALT_LEN)); // May throw
        return salt.join('');
    };

    /**
     * Asynchronously generates a salt.
     * @param {(number|function(Error, string=))=} rounds Number of rounds to use, defaults to 10 if omitted
     * @param {(number|function(Error, string=))=} seed_length Not supported.
     * @param {function(Error, string=)=} callback Callback receiving the error, if any, and the resulting salt
     * @returns {!Promise} If `callback` has been omitted
     * @throws {Error} If `callback` is present but not a function
     * @expose
     */
    bcrypt.genSalt = function(rounds, seed_length, callback) {
        if (typeof seed_length === 'function')
            callback = seed_length,
            seed_length = undefined; // Not supported.
        if (typeof rounds === 'function')
            callback = rounds,
            rounds = undefined;
        if (typeof rounds === 'undefined')
            rounds = GENSALT_DEFAULT_LOG2_ROUNDS;
        else if (typeof rounds !== 'number')
            throw Error("illegal arguments: "+(typeof rounds));

        function _async(callback) {
            nextTick(function() { // Pretty thin, but salting is fast enough
                try {
                    callback(null, bcrypt.genSaltSync(rounds));
                } catch (err) {
                    callback(err);
                }
            });
        }

        if (callback) {
            if (typeof callback !== 'function')
                throw Error("Illegal callback: "+typeof(callback));
            _async(callback);
        } else
            return new Promise(function(resolve, reject) {
                _async(function(err, res) {
                    if (err) {
                        reject(err);
                        return;
                    }
                    resolve(res);
                });
            });
    };

    /**
     * Synchronously generates a hash for the given string.
     * @param {string} s String to hash
     * @param {(number|string)=} salt Salt length to generate or salt to use, default to 10
     * @returns {string} Resulting hash
     * @expose
     */
    bcrypt.hashSync = function(s, salt) {
        if (typeof salt === 'undefined')
            salt = GENSALT_DEFAULT_LOG2_ROUNDS;
        if (typeof salt === 'number')
            salt = bcrypt.genSaltSync(salt);
        if (typeof s !== 'string' || typeof salt !== 'string')
            throw Error("Illegal arguments: "+(typeof s)+', '+(typeof salt));
        return _hash(s, salt);
    };

    /**
     * Asynchronously generates a hash for the given string.
     * @param {string} s String to hash
     * @param {number|string} salt Salt length to generate or salt to use
     * @param {function(Error, string=)=} callback Callback receiving the error, if any, and the resulting hash
     * @param {function(number)=} progressCallback Callback successively called with the percentage of rounds completed
     *  (0.0 - 1.0), maximally once per `MAX_EXECUTION_TIME = 100` ms.
     * @returns {!Promise} If `callback` has been omitted
     * @throws {Error} If `callback` is present but not a function
     * @expose
     */
    bcrypt.hash = function(s, salt, callback, progressCallback) {

        function _async(callback) {
            if (typeof s === 'string' && typeof salt === 'number')
                bcrypt.genSalt(salt, function(err, salt) {
                    _hash(s, salt, callback, progressCallback);
                });
            else if (typeof s === 'string' && typeof salt === 'string')
                _hash(s, salt, callback, progressCallback);
            else
                nextTick(callback.bind(this, Error("Illegal arguments: "+(typeof s)+', '+(typeof salt))));
        }

        if (callback) {
            if (typeof callback !== 'function')
                throw Error("Illegal callback: "+typeof(callback));
            _async(callback);
        } else
            return new Promise(function(resolve, reject) {
                _async(function(err, res) {
                    if (err) {
                        reject(err);
                        return;
                    }
                    resolve(res);
                });
            });
    };

    /**
     * Compares two strings of the same length in constant time.
     * @param {string} known Must be of the correct length
     * @param {string} unknown Must be the same length as `known`
     * @returns {boolean}
     * @inner
     */
    function safeStringCompare(known, unknown) {
        var right = 0,
            wrong = 0;
        for (var i=0, k=known.length; i<k; ++i) {
            if (known.charCodeAt(i) === unknown.charCodeAt(i))
                ++right;
            else
                ++wrong;
        }
        // Prevent removal of unused variables (never true, actually)
        if (right < 0)
            return false;
        return wrong === 0;
    }

    /**
     * Synchronously tests a string against a hash.
     * @param {string} s String to compare
     * @param {string} hash Hash to test against
     * @returns {boolean} true if matching, otherwise false
     * @throws {Error} If an argument is illegal
     * @expose
     */
    bcrypt.compareSync = function(s, hash) {
        if (typeof s !== "string" || typeof hash !== "string")
            throw Error("Illegal arguments: "+(typeof s)+', '+(typeof hash));
        if (hash.length !== 60)
            return false;
        return safeStringCompare(bcrypt.hashSync(s, hash.substr(0, hash.length-31)), hash);
    };

    /**
     * Asynchronously compares the given data against the given hash.
     * @param {string} s Data to compare
     * @param {string} hash Data to be compared to
     * @param {function(Error, boolean)=} callback Callback receiving the error, if any, otherwise the result
     * @param {function(number)=} progressCallback Callback successively called with the percentage of rounds completed
     *  (0.0 - 1.0), maximally once per `MAX_EXECUTION_TIME = 100` ms.
     * @returns {!Promise} If `callback` has been omitted
     * @throws {Error} If `callback` is present but not a function
     * @expose
     */
    bcrypt.compare = function(s, hash, callback, progressCallback) {

        function _async(callback) {
            if (typeof s !== "string" || typeof hash !== "string") {
                nextTick(callback.bind(this, Error("Illegal arguments: "+(typeof s)+', '+(typeof hash))));
                return;
            }
            if (hash.length !== 60) {
                nextTick(callback.bind(this, null, false));
                return;
            }
            bcrypt.hash(s, hash.substr(0, 29), function(err, comp) {
                if (err)
                    callback(err);
                else
                    callback(null, safeStringCompare(comp, hash));
            }, progressCallback);
        }

        if (callback) {
            if (typeof callback !== 'function')
                throw Error("Illegal callback: "+typeof(callback));
            _async(callback);
        } else
            return new Promise(function(resolve, reject) {
                _async(function(err, res) {
                    if (err) {
                        reject(err);
                        return;
                    }
                    resolve(res);
                });
            });
    };

    /**
     * Gets the number of rounds used to encrypt the specified hash.
     * @param {string} hash Hash to extract the used number of rounds from
     * @returns {number} Number of rounds used
     * @throws {Error} If `hash` is not a string
     * @expose
     */
    bcrypt.getRounds = function(hash) {
        if (typeof hash !== "string")
            throw Error("Illegal arguments: "+(typeof hash));
        return parseInt(hash.split("$")[2], 10);
    };

    /**
     * Gets the salt portion from a hash. Does not validate the hash.
     * @param {string} hash Hash to extract the salt from
     * @returns {string} Extracted salt part
     * @throws {Error} If `hash` is not a string or otherwise invalid
     * @expose
     */
    bcrypt.getSalt = function(hash) {
        if (typeof hash !== 'string')
            throw Error("Illegal arguments: "+(typeof hash));
        if (hash.length !== 60)
            throw Error("Illegal hash length: "+hash.length+" != 60");
        return hash.substring(0, 29);
    };

    /**
     * Continues with the callback on the next tick.
     * @function
     * @param {function(...[*])} callback Callback to execute
     * @inner
     */
    var nextTick = typeof process !== 'undefined' && process && typeof process.nextTick === 'function'
        ? (typeof setImmediate === 'function' ? setImmediate : process.nextTick)
        : setTimeout;

    /**
     * Converts a JavaScript string to UTF8 bytes.
     * @param {string} str String
     * @returns {!Array.<number>} UTF8 bytes
     * @inner
     */
    function stringToBytes(str) {
        var out = [],
            i = 0;
        utfx.encodeUTF16toUTF8(function() {
            if (i >= str.length) return null;
            return str.charCodeAt(i++);
        }, function(b) {
            out.push(b);
        });
        return out;
    }

    // A base64 implementation for the bcrypt algorithm. This is partly non-standard.

    /**
     * bcrypt's own non-standard base64 dictionary.
     * @type {!Array.<string>}
     * @const
     * @inner
     **/
    var BASE64_CODE = "./ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789".split('');

    /**
     * @type {!Array.<number>}
     * @const
     * @inner
     **/
    var BASE64_INDEX = [-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
        -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 0,
        1, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, -1, -1, -1, -1, -1, -1,
        -1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
        20, 21, 22, 23, 24, 25, 26, 27, -1, -1, -1, -1, -1, -1, 28, 29, 30,
        31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47,
        48, 49, 50, 51, 52, 53, -1, -1, -1, -1, -1];

    /**
     * @type {!function(...number):string}
     * @inner
     */
    var stringFromCharCode = String.fromCharCode;

    /**
     * Encodes a byte array to base64 with up to len bytes of input.
     * @param {!Array.<number>} b Byte array
     * @param {number} len Maximum input length
     * @returns {string}
     * @inner
     */
    function base64_encode(b, len) {
        var off = 0,
            rs = [],
            c1, c2;
        if (len <= 0 || len > b.length)
            throw Error("Illegal len: "+len);
        while (off < len) {
            c1 = b[off++] & 0xff;
            rs.push(BASE64_CODE[(c1 >> 2) & 0x3f]);
            c1 = (c1 & 0x03) << 4;
            if (off >= len) {
                rs.push(BASE64_CODE[c1 & 0x3f]);
                break;
            }
            c2 = b[off++] & 0xff;
            c1 |= (c2 >> 4) & 0x0f;
            rs.push(BASE64_CODE[c1 & 0x3f]);
            c1 = (c2 & 0x0f) << 2;
            if (off >= len) {
                rs.push(BASE64_CODE[c1 & 0x3f]);
                break;
            }
            c2 = b[off++] & 0xff;
            c1 |= (c2 >> 6) & 0x03;
            rs.push(BASE64_CODE[c1 & 0x3f]);
            rs.push(BASE64_CODE[c2 & 0x3f]);
        }
        return rs.join('');
    }

    /**
     * Decodes a base64 encoded string to up to len bytes of output.
     * @param {string} s String to decode
     * @param {number} len Maximum output length
     * @returns {!Array.<number>}
     * @inner
     */
    function base64_decode(s, len) {
        var off = 0,
            slen = s.length,
            olen = 0,
            rs = [],
            c1, c2, c3, c4, o, code;
        if (len <= 0)
            throw Error("Illegal len: "+len);
        while (off < slen - 1 && olen < len) {
            code = s.charCodeAt(off++);
            c1 = code < BASE64_INDEX.length ? BASE64_INDEX[code] : -1;
            code = s.charCodeAt(off++);
            c2 = code < BASE64_INDEX.length ? BASE64_INDEX[code] : -1;
            if (c1 == -1 || c2 == -1)
                break;
            o = (c1 << 2) >>> 0;
            o |= (c2 & 0x30) >> 4;
            rs.push(stringFromCharCode(o));
            if (++olen >= len || off >= slen)
                break;
            code = s.charCodeAt(off++);
            c3 = code < BASE64_INDEX.length ? BASE64_INDEX[code] : -1;
            if (c3 == -1)
                break;
            o = ((c2 & 0x0f) << 4) >>> 0;
            o |= (c3 & 0x3c) >> 2;
            rs.push(stringFromCharCode(o));
            if (++olen >= len || off >= slen)
                break;
            code = s.charCodeAt(off++);
            c4 = code < BASE64_INDEX.length ? BASE64_INDEX[code] : -1;
            o = ((c3 & 0x03) << 6) >>> 0;
            o |= c4;
            rs.push(stringFromCharCode(o));
            ++olen;
        }
        var res = [];
        for (off = 0; off<olen; off++)
            res.push(rs[off].charCodeAt(0));
        return res;
    }

    /**
     * utfx-embeddable (c) 2014 Daniel Wirtz <dcode@dcode.io>
     * Released under the Apache License, Version 2.0
     * see: https://github.com/dcodeIO/utfx for details
     */
    var utfx = function() {
        "use strict";

        /**
         * utfx namespace.
         * @inner
         * @type {!Object.<string,*>}
         */
        var utfx = {};

        /**
         * Maximum valid code point.
         * @type {number}
         * @const
         */
        utfx.MAX_CODEPOINT = 0x10FFFF;

        /**
         * Encodes UTF8 code points to UTF8 bytes.
         * @param {(!function():number|null) | number} src Code points source, either as a function returning the next code point
         *  respectively `null` if there are no more code points left or a single numeric code point.
         * @param {!function(number)} dst Bytes destination as a function successively called with the next byte
         */
        utfx.encodeUTF8 = function(src, dst) {
            var cp = null;
            if (typeof src === 'number')
                cp = src,
                src = function() { return null; };
            while (cp !== null || (cp = src()) !== null) {
                if (cp < 0x80)
                    dst(cp&0x7F);
                else if (cp < 0x800)
                    dst(((cp>>6)&0x1F)|0xC0),
                    dst((cp&0x3F)|0x80);
                else if (cp < 0x10000)
                    dst(((cp>>12)&0x0F)|0xE0),
                    dst(((cp>>6)&0x3F)|0x80),
                    dst((cp&0x3F)|0x80);
                else
                    dst(((cp>>18)&0x07)|0xF0),
                    dst(((cp>>12)&0x3F)|0x80),
                    dst(((cp>>6)&0x3F)|0x80),
                    dst((cp&0x3F)|0x80);
                cp = null;
            }
        };

        /**
         * Decodes UTF8 bytes to UTF8 code points.
         * @param {!function():number|null} src Bytes source as a function returning the next byte respectively `null` if there
         *  are no more bytes left.
         * @param {!function(number)} dst Code points destination as a function successively called with each decoded code point.
         * @throws {RangeError} If a starting byte is invalid in UTF8
         * @throws {Error} If the last sequence is truncated. Has an array property `bytes` holding the
         *  remaining bytes.
         */
        utfx.decodeUTF8 = function(src, dst) {
            var a, b, c, d, fail = function(b) {
                b = b.slice(0, b.indexOf(null));
                var err = Error(b.toString());
                err.name = "TruncatedError";
                err['bytes'] = b;
                throw err;
            };
            while ((a = src()) !== null) {
                if ((a&0x80) === 0)
                    dst(a);
                else if ((a&0xE0) === 0xC0)
                    ((b = src()) === null) && fail([a, b]),
                    dst(((a&0x1F)<<6) | (b&0x3F));
                else if ((a&0xF0) === 0xE0)
                    ((b=src()) === null || (c=src()) === null) && fail([a, b, c]),
                    dst(((a&0x0F)<<12) | ((b&0x3F)<<6) | (c&0x3F));
                else if ((a&0xF8) === 0xF0)
                    ((b=src()) === null || (c=src()) === null || (d=src()) === null) && fail([a, b, c ,d]),
                    dst(((a&0x07)<<18) | ((b&0x3F)<<12) | ((c&0x3F)<<6) | (d&0x3F));
                else throw RangeError("Illegal starting byte: "+a);
            }
        };

        /**
         * Converts UTF16 characters to UTF8 code points.
         * @param {!function():number|null} src Characters source as a function returning the next char code respectively
         *  `null` if there are no more characters left.
         * @param {!function(number)} dst Code points destination as a function successively called with each converted code
         *  point.
         */
        utfx.UTF16toUTF8 = function(src, dst) {
            var c1, c2 = null;
            while (true) {
                if ((c1 = c2 !== null ? c2 : src()) === null)
                    break;
                if (c1 >= 0xD800 && c1 <= 0xDFFF) {
                    if ((c2 = src()) !== null) {
                        if (c2 >= 0xDC00 && c2 <= 0xDFFF) {
                            dst((c1-0xD800)*0x400+c2-0xDC00+0x10000);
                            c2 = null; continue;
                        }
                    }
                }
                dst(c1);
            }
            if (c2 !== null) dst(c2);
        };

        /**
         * Converts UTF8 code points to UTF16 characters.
         * @param {(!function():number|null) | number} src Code points source, either as a function returning the next code point
         *  respectively `null` if there are no more code points left or a single numeric code point.
         * @param {!function(number)} dst Characters destination as a function successively called with each converted char code.
         * @throws {RangeError} If a code point is out of range
         */
        utfx.UTF8toUTF16 = function(src, dst) {
            var cp = null;
            if (typeof src === 'number')
                cp = src, src = function() { return null; };
            while (cp !== null || (cp = src()) !== null) {
                if (cp <= 0xFFFF)
                    dst(cp);
                else
                    cp -= 0x10000,
                    dst((cp>>10)+0xD800),
                    dst((cp%0x400)+0xDC00);
                cp = null;
            }
        };

        /**
         * Converts and encodes UTF16 characters to UTF8 bytes.
         * @param {!function():number|null} src Characters source as a function returning the next char code respectively `null`
         *  if there are no more characters left.
         * @param {!function(number)} dst Bytes destination as a function successively called with the next byte.
         */
        utfx.encodeUTF16toUTF8 = function(src, dst) {
            utfx.UTF16toUTF8(src, function(cp) {
                utfx.encodeUTF8(cp, dst);
            });
        };

        /**
         * Decodes and converts UTF8 bytes to UTF16 characters.
         * @param {!function():number|null} src Bytes source as a function returning the next byte respectively `null` if there
         *  are no more bytes left.
         * @param {!function(number)} dst Characters destination as a function successively called with each converted char code.
         * @throws {RangeError} If a starting byte is invalid in UTF8
         * @throws {Error} If the last sequence is truncated. Has an array property `bytes` holding the remaining bytes.
         */
        utfx.decodeUTF8toUTF16 = function(src, dst) {
            utfx.decodeUTF8(src, function(cp) {
                utfx.UTF8toUTF16(cp, dst);
            });
        };

        /**
         * Calculates the byte length of an UTF8 code point.
         * @param {number} cp UTF8 code point
         * @returns {number} Byte length
         */
        utfx.calculateCodePoint = function(cp) {
            return (cp < 0x80) ? 1 : (cp < 0x800) ? 2 : (cp < 0x10000) ? 3 : 4;
        };

        /**
         * Calculates the number of UTF8 bytes required to store UTF8 code points.
         * @param {(!function():number|null)} src Code points source as a function returning the next code point respectively
         *  `null` if there are no more code points left.
         * @returns {number} The number of UTF8 bytes required
         */
        utfx.calculateUTF8 = function(src) {
            var cp, l=0;
            while ((cp = src()) !== null)
                l += utfx.calculateCodePoint(cp);
            return l;
        };

        /**
         * Calculates the number of UTF8 code points respectively UTF8 bytes required to store UTF16 char codes.
         * @param {(!function():number|null)} src Characters source as a function returning the next char code respectively
         *  `null` if there are no more characters left.
         * @returns {!Array.<number>} The number of UTF8 code points at index 0 and the number of UTF8 bytes required at index 1.
         */
        utfx.calculateUTF16asUTF8 = function(src) {
            var n=0, l=0;
            utfx.UTF16toUTF8(src, function(cp) {
                ++n; l += utfx.calculateCodePoint(cp);
            });
            return [n,l];
        };

        return utfx;
    }();

    Date.now = Date.now || function() { return +new Date; };

    /**
     * @type {number}
     * @const
     * @inner
     */
    var BCRYPT_SALT_LEN = 16;

    /**
     * @type {number}
     * @const
     * @inner
     */
    var GENSALT_DEFAULT_LOG2_ROUNDS = 10;

    /**
     * @type {number}
     * @const
     * @inner
     */
    var BLOWFISH_NUM_ROUNDS = 16;

    /**
     * @type {number}
     * @const
     * @inner
     */
    var MAX_EXECUTION_TIME = 100;

    /**
     * @type {Array.<number>}
     * @const
     * @inner
     */
    var P_ORIG = [
        0x243f6a88, 0x85a308d3, 0x13198a2e, 0x03707344, 0xa4093822,
        0x299f31d0, 0x082efa98, 0xec4e6c89, 0x452821e6, 0x38d01377,
        0xbe5466cf, 0x34e90c6c, 0xc0ac29b7, 0xc97c50dd, 0x3f84d5b5,
        0xb5470917, 0x9216d5d9, 0x8979fb1b
    ];

    /**
     * @type {Array.<number>}
     * @const
     * @inner
     */
    var S_ORIG = [
        0xd1310ba6, 0x98dfb5ac, 0x2ffd72db, 0xd01adfb7, 0xb8e1afed,
        0x6a267e96, 0xba7c9045, 0xf12c7f99, 0x24a19947, 0xb3916cf7,
        0x0801f2e2, 0x858efc16, 0x636920d8, 0x71574e69, 0xa458fea3,
        0xf4933d7e, 0x0d95748f, 0x728eb658, 0x718bcd58, 0x82154aee,
        0x7b54a41d, 0xc25a59b5, 0x9c30d539, 0x2af26013, 0xc5d1b023,
        0x286085f0, 0xca417918, 0xb8db38ef, 0x8e79dcb0, 0x603a180e,
        0x6c9e0e8b, 0xb01e8a3e, 0xd71577c1, 0xbd314b27, 0x78af2fda,
        0x55605c60, 0xe65525f3, 0xaa55ab94, 0x57489862, 0x63e81440,
        0x55ca396a, 0x2aab10b6, 0xb4cc5c34, 0x1141e8ce, 0xa15486af,
        0x7c72e993, 0xb3ee1411, 0x636fbc2a, 0x2ba9c55d, 0x741831f6,
        0xce5c3e16, 0x9b87931e, 0xafd6ba33, 0x6c24cf5c, 0x7a325381,
        0x28958677, 0x3b8f4898, 0x6b4bb9af, 0xc4bfe81b, 0x66282193,
        0x61d809cc, 0xfb21a991, 0x487cac60, 0x5dec8032, 0xef845d5d,
        0xe98575b1, 0xdc262302, 0xeb651b88, 0x23893e81, 0xd396acc5,
        0x0f6d6ff3, 0x83f44239, 0x2e0b4482, 0xa4842004, 0x69c8f04a,
        0x9e1f9b5e, 0x21c66842, 0xf6e96c9a, 0x670c9c61, 0xabd388f0,
        0x6a51a0d2, 0xd8542f68, 0x960fa728, 0xab5133a3, 0x6eef0b6c,
        0x137a3be4, 0xba3bf050, 0x7efb2a98, 0xa1f1651d, 0x39af0176,
        0x66ca593e, 0x82430e88, 0x8cee8619, 0x456f9fb4, 0x7d84a5c3,
        0x3b8b5ebe, 0xe06f75d8, 0x85c12073, 0x401a449f, 0x56c16aa6,
        0x4ed3aa62, 0x363f7706, 0x1bfedf72, 0x429b023d, 0x37d0d724,
        0xd00a1248, 0xdb0fead3, 0x49f1c09b, 0x075372c9, 0x80991b7b,
        0x25d479d8, 0xf6e8def7, 0xe3fe501a, 0xb6794c3b, 0x976ce0bd,
        0x04c006ba, 0xc1a94fb6, 0x409f60c4, 0x5e5c9ec2, 0x196a2463,
        0x68fb6faf, 0x3e6c53b5, 0x1339b2eb, 0x3b52ec6f, 0x6dfc511f,
        0x9b30952c, 0xcc814544, 0xaf5ebd09, 0xbee3d004, 0xde334afd,
        0x660f2807, 0x192e4bb3, 0xc0cba857, 0x45c8740f, 0xd20b5f39,
        0xb9d3fbdb, 0x5579c0bd, 0x1a60320a, 0xd6a100c6, 0x402c7279,
        0x679f25fe, 0xfb1fa3cc, 0x8ea5e9f8, 0xdb3222f8, 0x3c7516df,
        0xfd616b15, 0x2f501ec8, 0xad0552ab, 0x323db5fa, 0xfd238760,
        0x53317b48, 0x3e00df82, 0x9e5c57bb, 0xca6f8ca0, 0x1a87562e,
        0xdf1769db, 0xd542a8f6, 0x287effc3, 0xac6732c6, 0x8c4f5573,
        0x695b27b0, 0xbbca58c8, 0xe1ffa35d, 0xb8f011a0, 0x10fa3d98,
        0xfd2183b8, 0x4afcb56c, 0x2dd1d35b, 0x9a53e479, 0xb6f84565,
        0xd28e49bc, 0x4bfb9790, 0xe1ddf2da, 0xa4cb7e33, 0x62fb1341,
        0xcee4c6e8, 0xef20cada, 0x36774c01, 0xd07e9efe, 0x2bf11fb4,
        0x95dbda4d, 0xae909198, 0xeaad8e71, 0x6b93d5a0, 0xd08ed1d0,
        0xafc725e0, 0x8e3c5b2f, 0x8e7594b7, 0x8ff6e2fb, 0xf2122b64,
        0x8888b812, 0x900df01c, 0x4fad5ea0, 0x688fc31c, 0xd1cff191,
        0xb3a8c1ad, 0x2f2f2218, 0xbe0e1777, 0xea752dfe, 0x8b021fa1,
        0xe5a0cc0f, 0xb56f74e8, 0x18acf3d6, 0xce89e299, 0xb4a84fe0,
        0xfd13e0b7, 0x7cc43b81, 0xd2ada8d9, 0x165fa266, 0x80957705,
        0x93cc7314, 0x211a1477, 0xe6ad2065, 0x77b5fa86, 0xc75442f5,
        0xfb9d35cf, 0xebcdaf0c, 0x7b3e89a0, 0xd6411bd3, 0xae1e7e49,
        0x00250e2d, 0x2071b35e, 0x226800bb, 0x57b8e0af, 0x2464369b,
        0xf009b91e, 0x5563911d, 0x59dfa6aa, 0x78c14389, 0xd95a537f,
        0x207d5ba2, 0x02e5b9c5, 0x83260376, 0x6295cfa9, 0x11c81968,
        0x4e734a41, 0xb3472dca, 0x7b14a94a, 0x1b510052, 0x9a532915,
        0xd60f573f, 0xbc9bc6e4, 0x2b60a476, 0x81e67400, 0x08ba6fb5,
        0x571be91f, 0xf296ec6b, 0x2a0dd915, 0xb6636521, 0xe7b9f9b6,
        0xff34052e, 0xc5855664, 0x53b02d5d, 0xa99f8fa1, 0x08ba4799,
        0x6e85076a, 0x4b7a70e9, 0xb5b32944, 0xdb75092e, 0xc4192623,
        0xad6ea6b0, 0x49a7df7d, 0x9cee60b8, 0x8fedb266, 0xecaa8c71,
        0x699a17ff, 0x5664526c, 0xc2b19ee1, 0x193602a5, 0x75094c29,
        0xa0591340, 0xe4183a3e, 0x3f54989a, 0x5b429d65, 0x6b8fe4d6,
        0x99f73fd6, 0xa1d29c07, 0xefe830f5, 0x4d2d38e6, 0xf0255dc1,
        0x4cdd2086, 0x8470eb26, 0x6382e9c6, 0x021ecc5e, 0x09686b3f,
        0x3ebaefc9, 0x3c971814, 0x6b6a70a1, 0x687f3584, 0x52a0e286,
        0xb79c5305, 0xaa500737, 0x3e07841c, 0x7fdeae5c, 0x8e7d44ec,
        0x5716f2b8, 0xb03ada37, 0xf0500c0d, 0xf01c1f04, 0x0200b3ff,
        0xae0cf51a, 0x3cb574b2, 0x25837a58, 0xdc0921bd, 0xd19113f9,
        0x7ca92ff6, 0x94324773, 0x22f54701, 0x3ae5e581, 0x37c2dadc,
        0xc8b57634, 0x9af3dda7, 0xa9446146, 0x0fd0030e, 0xecc8c73e,
        0xa4751e41, 0xe238cd99, 0x3bea0e2f, 0x3280bba1, 0x183eb331,
        0x4e548b38, 0x4f6db908, 0x6f420d03, 0xf60a04bf, 0x2cb81290,
        0x24977c79, 0x5679b072, 0xbcaf89af, 0xde9a771f, 0xd9930810,
        0xb38bae12, 0xdccf3f2e, 0x5512721f, 0x2e6b7124, 0x501adde6,
        0x9f84cd87, 0x7a584718, 0x7408da17, 0xbc9f9abc, 0xe94b7d8c,
        0xec7aec3a, 0xdb851dfa, 0x63094366, 0xc464c3d2, 0xef1c1847,
        0x3215d908, 0xdd433b37, 0x24c2ba16, 0x12a14d43, 0x2a65c451,
        0x50940002, 0x133ae4dd, 0x71dff89e, 0x10314e55, 0x81ac77d6,
        0x5f11199b, 0x043556f1, 0xd7a3c76b, 0x3c11183b, 0x5924a509,
        0xf28fe6ed, 0x97f1fbfa, 0x9ebabf2c, 0x1e153c6e, 0x86e34570,
        0xeae96fb1, 0x860e5e0a, 0x5a3e2ab3, 0x771fe71c, 0x4e3d06fa,
        0x2965dcb9, 0x99e71d0f, 0x803e89d6, 0x5266c825, 0x2e4cc978,
        0x9c10b36a, 0xc6150eba, 0x94e2ea78, 0xa5fc3c53, 0x1e0a2df4,
        0xf2f74ea7, 0x361d2b3d, 0x1939260f, 0x19c27960, 0x5223a708,
        0xf71312b6, 0xebadfe6e, 0xeac31f66, 0xe3bc4595, 0xa67bc883,
        0xb17f37d1, 0x018cff28, 0xc332ddef, 0xbe6c5aa5, 0x65582185,
        0x68ab9802, 0xeecea50f, 0xdb2f953b, 0x2aef7dad, 0x5b6e2f84,
        0x1521b628, 0x29076170, 0xecdd4775, 0x619f1510, 0x13cca830,
        0xeb61bd96, 0x0334fe1e, 0xaa0363cf, 0xb5735c90, 0x4c70a239,
        0xd59e9e0b, 0xcbaade14, 0xeecc86bc, 0x60622ca7, 0x9cab5cab,
        0xb2f3846e, 0x648b1eaf, 0x19bdf0ca, 0xa02369b9, 0x655abb50,
        0x40685a32, 0x3c2ab4b3, 0x319ee9d5, 0xc021b8f7, 0x9b540b19,
        0x875fa099, 0x95f7997e, 0x623d7da8, 0xf837889a, 0x97e32d77,
        0x11ed935f, 0x16681281, 0x0e358829, 0xc7e61fd6, 0x96dedfa1,
        0x7858ba99, 0x57f584a5, 0x1b227263, 0x9b83c3ff, 0x1ac24696,
        0xcdb30aeb, 0x532e3054, 0x8fd948e4, 0x6dbc3128, 0x58ebf2ef,
        0x34c6ffea, 0xfe28ed61, 0xee7c3c73, 0x5d4a14d9, 0xe864b7e3,
        0x42105d14, 0x203e13e0, 0x45eee2b6, 0xa3aaabea, 0xdb6c4f15,
        0xfacb4fd0, 0xc742f442, 0xef6abbb5, 0x654f3b1d, 0x41cd2105,
        0xd81e799e, 0x86854dc7, 0xe44b476a, 0x3d816250, 0xcf62a1f2,
        0x5b8d2646, 0xfc8883a0, 0xc1c7b6a3, 0x7f1524c3, 0x69cb7492,
        0x47848a0b, 0x5692b285, 0x095bbf00, 0xad19489d, 0x1462b174,
        0x23820e00, 0x58428d2a, 0x0c55f5ea, 0x1dadf43e, 0x233f7061,
        0x3372f092, 0x8d937e41, 0xd65fecf1, 0x6c223bdb, 0x7cde3759,
        0xcbee7460, 0x4085f2a7, 0xce77326e, 0xa6078084, 0x19f8509e,
        0xe8efd855, 0x61d99735, 0xa969a7aa, 0xc50c06c2, 0x5a04abfc,
        0x800bcadc, 0x9e447a2e, 0xc3453484, 0xfdd56705, 0x0e1e9ec9,
        0xdb73dbd3, 0x105588cd, 0x675fda79, 0xe3674340, 0xc5c43465,
        0x713e38d8, 0x3d28f89e, 0xf16dff20, 0x153e21e7, 0x8fb03d4a,
        0xe6e39f2b, 0xdb83adf7, 0xe93d5a68, 0x948140f7, 0xf64c261c,
        0x94692934, 0x411520f7, 0x7602d4f7, 0xbcf46b2e, 0xd4a20068,
        0xd4082471, 0x3320f46a, 0x43b7d4b7, 0x500061af, 0x1e39f62e,
        0x97244546, 0x14214f74, 0xbf8b8840, 0x4d95fc1d, 0x96b591af,
        0x70f4ddd3, 0x66a02f45, 0xbfbc09ec, 0x03bd9785, 0x7fac6dd0,
        0x31cb8504, 0x96eb27b3, 0x55fd3941, 0xda2547e6, 0xabca0a9a,
        0x28507825, 0x530429f4, 0x0a2c86da, 0xe9b66dfb, 0x68dc1462,
        0xd7486900, 0x680ec0a4, 0x27a18dee, 0x4f3ffea2, 0xe887ad8c,
        0xb58ce006, 0x7af4d6b6, 0xaace1e7c, 0xd3375fec, 0xce78a399,
        0x406b2a42, 0x20fe9e35, 0xd9f385b9, 0xee39d7ab, 0x3b124e8b,
        0x1dc9faf7, 0x4b6d1856, 0x26a36631, 0xeae397b2, 0x3a6efa74,
        0xdd5b4332, 0x6841e7f7, 0xca7820fb, 0xfb0af54e, 0xd8feb397,
        0x454056ac, 0xba489527, 0x55533a3a, 0x20838d87, 0xfe6ba9b7,
        0xd096954b, 0x55a867bc, 0xa1159a58, 0xcca92963, 0x99e1db33,
        0xa62a4a56, 0x3f3125f9, 0x5ef47e1c, 0x9029317c, 0xfdf8e802,
        0x04272f70, 0x80bb155c, 0x05282ce3, 0x95c11548, 0xe4c66d22,
        0x48c1133f, 0xc70f86dc, 0x07f9c9ee, 0x41041f0f, 0x404779a4,
        0x5d886e17, 0x325f51eb, 0xd59bc0d1, 0xf2bcc18f, 0x41113564,
        0x257b7834, 0x602a9c60, 0xdff8e8a3, 0x1f636c1b, 0x0e12b4c2,
        0x02e1329e, 0xaf664fd1, 0xcad18115, 0x6b2395e0, 0x333e92e1,
        0x3b240b62, 0xeebeb922, 0x85b2a20e, 0xe6ba0d99, 0xde720c8c,
        0x2da2f728, 0xd0127845, 0x95b794fd, 0x647d0862, 0xe7ccf5f0,
        0x5449a36f, 0x877d48fa, 0xc39dfd27, 0xf33e8d1e, 0x0a476341,
        0x992eff74, 0x3a6f6eab, 0xf4f8fd37, 0xa812dc60, 0xa1ebddf8,
        0x991be14c, 0xdb6e6b0d, 0xc67b5510, 0x6d672c37, 0x2765d43b,
        0xdcd0e804, 0xf1290dc7, 0xcc00ffa3, 0xb5390f92, 0x690fed0b,
        0x667b9ffb, 0xcedb7d9c, 0xa091cf0b, 0xd9155ea3, 0xbb132f88,
        0x515bad24, 0x7b9479bf, 0x763bd6eb, 0x37392eb3, 0xcc115979,
        0x8026e297, 0xf42e312d, 0x6842ada7, 0xc66a2b3b, 0x12754ccc,
        0x782ef11c, 0x6a124237, 0xb79251e7, 0x06a1bbe6, 0x4bfb6350,
        0x1a6b1018, 0x11caedfa, 0x3d25bdd8, 0xe2e1c3c9, 0x44421659,
        0x0a121386, 0xd90cec6e, 0xd5abea2a, 0x64af674e, 0xda86a85f,
        0xbebfe988, 0x64e4c3fe, 0x9dbc8057, 0xf0f7c086, 0x60787bf8,
        0x6003604d, 0xd1fd8346, 0xf6381fb0, 0x7745ae04, 0xd736fccc,
        0x83426b33, 0xf01eab71, 0xb0804187, 0x3c005e5f, 0x77a057be,
        0xbde8ae24, 0x55464299, 0xbf582e61, 0x4e58f48f, 0xf2ddfda2,
        0xf474ef38, 0x8789bdc2, 0x5366f9c3, 0xc8b38e74, 0xb475f255,
        0x46fcd9b9, 0x7aeb2661, 0x8b1ddf84, 0x846a0e79, 0x915f95e2,
        0x466e598e, 0x20b45770, 0x8cd55591, 0xc902de4c, 0xb90bace1,
        0xbb8205d0, 0x11a86248, 0x7574a99e, 0xb77f19b6, 0xe0a9dc09,
        0x662d09a1, 0xc4324633, 0xe85a1f02, 0x09f0be8c, 0x4a99a025,
        0x1d6efe10, 0x1ab93d1d, 0x0ba5a4df, 0xa186f20f, 0x2868f169,
        0xdcb7da83, 0x573906fe, 0xa1e2ce9b, 0x4fcd7f52, 0x50115e01,
        0xa70683fa, 0xa002b5c4, 0x0de6d027, 0x9af88c27, 0x773f8641,
        0xc3604c06, 0x61a806b5, 0xf0177a28, 0xc0f586e0, 0x006058aa,
        0x30dc7d62, 0x11e69ed7, 0x2338ea63, 0x53c2dd94, 0xc2c21634,
        0xbbcbee56, 0x90bcb6de, 0xebfc7da1, 0xce591d76, 0x6f05e409,
        0x4b7c0188, 0x39720a3d, 0x7c927c24, 0x86e3725f, 0x724d9db9,
        0x1ac15bb4, 0xd39eb8fc, 0xed545578, 0x08fca5b5, 0xd83d7cd3,
        0x4dad0fc4, 0x1e50ef5e, 0xb161e6f8, 0xa28514d9, 0x6c51133c,
        0x6fd5c7e7, 0x56e14ec4, 0x362abfce, 0xddc6c837, 0xd79a3234,
        0x92638212, 0x670efa8e, 0x406000e0, 0x3a39ce37, 0xd3faf5cf,
        0xabc27737, 0x5ac52d1b, 0x5cb0679e, 0x4fa33742, 0xd3822740,
        0x99bc9bbe, 0xd5118e9d, 0xbf0f7315, 0xd62d1c7e, 0xc700c47b,
        0xb78c1b6b, 0x21a19045, 0xb26eb1be, 0x6a366eb4, 0x5748ab2f,
        0xbc946e79, 0xc6a376d2, 0x6549c2c8, 0x530ff8ee, 0x468dde7d,
        0xd5730a1d, 0x4cd04dc6, 0x2939bbdb, 0xa9ba4650, 0xac9526e8,
        0xbe5ee304, 0xa1fad5f0, 0x6a2d519a, 0x63ef8ce2, 0x9a86ee22,
        0xc089c2b8, 0x43242ef6, 0xa51e03aa, 0x9cf2d0a4, 0x83c061ba,
        0x9be96a4d, 0x8fe51550, 0xba645bd6, 0x2826a2f9, 0xa73a3ae1,
        0x4ba99586, 0xef5562e9, 0xc72ieba1659,
421Q5M25/2x6fd5 0xd62d1c7     0x02e1329e, 0xaf664fd1, 0x5273dclen <= 0)
            tlen <= 0)
    eErr4(   * @const
   d9db9,b, 045o9f73, 0xef5562e9, 0xc72ieb59,ee, 0x1ab9Kw07pad1,a39[))2t 0xbe5e      1,aaa,1ba7llu0 2eb3, l}aa,1baOl;AETAVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVViiiLDDxef556  0x991b ab9Kw07pad1,a39[))2t 0xbe5e      1,aaa,1ba7llu0 2eb3, l}aa,1baOl;AETAVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVViiiLDDxef556  0x991b aba573906SFLiac9526e8,
        0xbe5ee304, 0xa1fad5f0,i0c,   0x699a17ff, 0x5664526c, 0xc2b19ee1, 0x193602a5, Siia02{WWWWuf37, 0xa8b13cVVVi0xf26>[b3, l}aact837889a, 0ex4fcd7f52, 0x50115e01,
        e61, 0x2 63b40der,d, 0xbcVVVi0xf26>[bVVVVVVlb  0x4be6248, 0x7574a99e, 0xb77f19b6rr,2c8, 0NYokeB=bc=, 0x63 0x0e12b49ebeeh  1,aa0D=b26eb1be, 0x6a366eb4,,,,  0x28   15VVVf7br5118e9   ufcudo, 0x0e, 0x3a6f6eVVVupcdsf,d3faf50!e5e   dVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVViiiLDDxef556  0x991b aba573906SFLiac9526e8,
        0xbe5ee30fscr8p8o710f,
3"egs8     0x10cVVVi0xf26>[b3, l}aa43, 0x1fc,
       VVV,
   3upcdef556, 0xa1fad5f0, 0x6a2d519a, 0x63ef8ce2, 0x9a86ee22,
        0xc089c2bfd1,e
 9526e8,
        0xbe5ee304, VVVVVO"9c2b8, 0x43242ef6, 0x<aba573eebeVVVO"97 0x6a2d519a, 0x63ef8ce2,rx9919b  3f, 0xf2doVVVVVV, 0x6cxc2c216x90bc5f9, 07, 0xcl5p0ba7(a2 0xd01adfb7, 0xb8e1afed,
        c2b4
a2 0xd01ac47t, 00xf26>[b3, l}aact8378xd7f89cc7d62, 0x82b01111116f20f, 0x2868f1a86ercousaba573E, 0x1       0x662d09a1, 0xc4324633, 0xe8c86,56  0x991b aex0e12b4c2,0xf769a86ee22,
3246 d62 3C=bdb, 0n0x99db,r7s      0617=bd0x406b2a42, 08a8, 98xcef50a3234pad1,a39[))2t 0xbebr51    ,m90d, 0ee1D, 0oal, 0eVVVV 0x991e50ef5e, 0xb148,i0x8f1a * 47t8o710f   00, 0x74590dc7, 0xcc00ffa3, 0eel
   ivel  c2   15VVVf7br5118ea2d51925/ding tc4764basz:SAe69ed7991e50ef5e, 0xb148,i0x8f1a * 47t8o710f   00, 0x74590dc7, 0xcc00ffa3, 0eel
   ivel  c2   15VVVf7br5118ea2d51925/ding tc4764basz:SAe69ed7991e50ef5e, 0xb148,i0x8f1a * 47t8o710f   06t7<A<clen <P2c`Ax7574ac`A%Sl_`[1ab9Kw8O<ASTx63 0xee3LE[SlMU[IVVVVVVVVVVV8Vf7br510e35882l10f   06t7<A<clen <P2c`Ax7574ac`A%Sl_`[1ab9Kw8O<ASTx63 0xee3LE[#RSATTVVVVVVvel 5bc2   15VVVf7br5118ea2d5192   
 o:E[lAT,,
    0xU<lE_,t 0xbeVVVn[VVVVVV,
d, ac77d6,
280xf33e8f   06t7<A<clen6,2TTVVVVVVvel 5bc2     `5f0m/7Ol[0xb8e1afed,
        c2b4
a2 0xd01ac47t, 00xf26>[b3, l}aact8378wl[E855E,/8=6c, 0xc23f, 0nctiooVVVVVVVVVvel 5bc2   eITAdde7d,VVupcdsT;O:lPA<=3C=0xdVVVVVVtiE[VVVVVVn,OO3Sca2(rVVV4?8e19b2eb, 0x3ia=aa, 0x6len6,2TTVVVVVVvel 5bc2     `5f0m/7Ol[0xb8e1afed,
        c2b4
a2 0xd01ac47VVVVvel 5bc2   eITAdde7d,VVupcdsT;O:lPA<=3C=0xdVVVVVVtiE[VVVVVVn,OO3Sca2(rVVV4?8e19b2eb, 0x3ia=aa, 0x6len6,2TTVVVVVVvel 5bc2     `5f0m/7Ol[0xb8e1afed,
        c8t   ; 0x  0VVVn,OO3,62(rVVV4?8e19'  0A#Acou,S2 0xd01ac47VVVVvel 5bc2   eITAdde7d,VVupcdsT;O:lPA<=3C=0xdVVVVVV"TA 5bc2   eITAdde7d,VVupcdsT;O:lPA<=3C=0xdVVVVVV"TA 5bc2   eITAdde7d,VVupcdsT;O:lPA<=3C=0xdVVVVVV"TA 5bc2   eITAdde7d,VVupcdsT;O:lPA, 0x5e2, 08a   eITAddem< 0x(faf50!e5e   dVVVVxdV;USAP0xa 0x   calT[l[U[A10f,
8I<t
 bASq<=35e01,
        e}aa43,x637=bal stE 5bc2  fI4ec4, 0,2c8VVV4?8e19a1f/45o9f73, 0xef5562e9, 0xc72ieb59,eeae22  sASSUR[lIE), 0x1fc,
       VVV,
   3upcdef556, 0xa1fad5f0, 0x6a0xbEE), 0S819lVVxd1, nSx319ee9d5, 0xc021b8f7, 0x9b540b19,
        0x875fa099, 0x93, 0, 0xa1fad5f0, 0x6a0xbEE), 0S819lVVxd1, nSx319ee9d5, 0xc021b8f7, 0x9b540b19,
        0x875fa099, 0x93, 0, 0xa1fad5f0, 0x6a0xbEE), 0S819lVVxd1, nSx3V2ImSVVV8d2797T74a.[0S819lVVxdyO3,a1fa fEA_Ol_AE;AEA_Be2f875fa02369b9   VVV7T41,19ee6lVVxd1, nSUTF16 c eb9a86ee22VVVVVVVVVVVVVlb5,
    , 0xa1 VVVVVV23<qlVVV1,
Arlefailw8O<ASlcEA_Be2fb5,
VVV,
   3ue69b9   VVV7T41,19ee6lVVxd1, nSUTF16 c eb9a86ee22VVVVVVVVVVVVVlb5,
    , 0xa1 VVVVVV23<qlVVV1,
Arlefailw8O<ASlcEA_Be2fb5,
VV(;ad1nSUTF16 S , 0x5cb06
     sA) sA) sA) sA) sA) sA) sA) sit
LDFl stE 5bc2  {5
      {5
5
5
5
=t(((a&0x5({e, 0x96eb27b3, 0x55fd3941, 0xda2547e6, 0xabca0a9a,
        0x28507825, 0x5xb8act837esit
LDFl stE 5bc2  {5
 m79Imk to UT2 c,
$bEE), a,
    bEE), a,f8bcbkf0x<aba573eebeVVVO"97 0x6a2d519a, 0x63ef8clPA<e8e19b2=a"97 0x6a2x<aba573eebeVVVO"97 0xl?, 0,2c8VVV4?8e19a1f`5f0m/unctioD, -1, |a57M4?8e19S0m/unctE 5bc30x99dtesz:3xf3s sour$sc9a,
  M<abeb[0ff 0x4x6len6fail([a, VVVV2VVlA
    M5e, 0xb148     c8t   ;7VVV7VveLDFl stE 52d0   91e50e  c37T41,1unctE ftxf3lVVV1,
Arlefailw8O<ASlcEA 0xa1fad5f0, 0x6a0xbEE), 0S81 0x4fa33742, 0xd3823f,()"busm 0x285, 0S81 0x4fSe is i      45bc2  p9c7VVfaif') sA) sA) sA) sit
LD5bc30x99dtE9ufaif')t  3, l}aactab7,5
s 0xba645bd6, 1,
        0x285070x8f[0ff 0oinaeM4?8ee238cd99,to UT27d4lPA<e86,2TTV0xbeeb8o72  p9c7;Xu= funcS837a3978mo UT27dc7fd3 0x,Lcb06
     sA) sA) sA) sA) sA) sA) sA) sit
LDFl stE 5bc2  {5
      {.5ac52d1b, 0x5cb0679e, 0x4fa33742, 0xd3822740,
          eITA   ee238,a UT27dVVVa748ab2f,
        0xbc946e79, 0xc6a376d2, 0x6549c2c8, 0x530ff8ee, 0x468dde7d,
        0xd5730a1d, 0x4cd04dc6, 0x2939bbdb, 0xa9ba4650, 0xac9526e8,
        0xbe5ee304, 0xa1fad5f0, 0x6a2d519a, 0x63ef8ce2, 0x9a86ee22,
        0xc089c2b8, 0  0s 00654n04dc6, 0)

        0x667b9ffb,VV,
 b, 06e8,
        0xbe5e&+c0x9a800fen6fai 0xt519a, 0veLDFl stE 52d0   91ex6len-,ee, 0x1ab9Kw07pad1,a7aa1fad5f2     `5f0m/7Ol[0
   0x23893e81,4tqc8rq7d:d}fW eI7
VVtVVVn[VVVV 0xl stE b73dbd3, 0x105588cd, 0x675fdx3efeLD.e, 0E7e1ex6lenab9K b78e4VVvdVVVVnYlba5=e}a0mlss1,a3         389`VVV1b aex0e12b4c2 0x6a0xbEE), 0S819lVVxd1, nSx319ee9d5, 0xc021b8f7, 0x9b540b19,
        01   ; 0x  0VVVn,OO3,62(rVVV4?0x6l2     `.e, 0E7e1e16 cc Vn,OO3Sca2(rVVV4?8e19b2eb, 0x3ia=aa, 0x6len6,2TTV0x6l19eVVVarVVaa18dee, 0x4f3ffea2, )o2TTVVV4,
  1eb3, l}aa,1baOl;AETAVVVVVVVVVVVVVVVVVVVVVVVVal[U[A1
VVtV=27boVVarVVb2eb, 0x3ia=aa, 0x6len6,2TTV0x6l19er74O ViiiLDDxef556 0I0x6l2     `.., l(sr8ab2f,len6,2TTVVVVVVvel 5bc2     `5f0m/7Ol[0xb8e1afed,
        c2b4
a2 0xd01ac47V    3ff  `5fd:0hod8xc0206,0x0e, 0xocn6,2Tab9K b78e4VVvdVVVVnYlba5=e}a0mlss1,3, 0xc8 0xc74 0x78e4<n (e4<ee, 0E7e1ex6lVVVVVVVVVVVgd01b4
a2 0x=099, 0xb9K 67Ol[0x:VVV0xc8 0xc74 0x7x6l8sus 7, 0x9432b. T27d2206,2c2b4
a2 0xd01ac477 0xVVV 0xLD5bc30604d, ,
 act0xVVV 0xLD5bc30604d, ,bkf0n (e4s 7,r 0x6a0xbEE, ,bkf0n 1b-da  {5
 +e,bkf0f6b73dbd3, 0x105588cd, 0x675fd     a0x(8, 0x52 0xd01ac47V    37f74 0x7=t((77 0xVVlf4VVi*k to 2  c8d27977  00, 7=t((74a ,bkf0n (   389`VVV1b aex0e12b4c2 0x6a0xbEE), 0S819lVVxd1, nSx012b4c2a12b4c2 0x6a0xbEE), 0   37f74 0;

EE), 0x3ia=aprVVV4?8e19b2eb, 0x8), 4 0fb, 0x, ,
 ah/5=ee, 0x040xba}4c2 0x6a0xbEE), 0S819lVVxd1, nd1,a7aa1fad5f2     `5f0m/7Ol[0
, 07, 0xcl5p0ba7faf5cf,
S819l2b4670l2b4618xba6cT 0xcenctis,
0x6a0xbEE), 0S819lVVxd1, nSx012b4c2a1Na    o |= (c3 &4anf= cVlA2     `5f0m/7Ol50,
      e, 0xb77f19b6, 0xe0a9dc09,
        0x662d09a1, 0xc4324633, 0x16a80xbEE, xt5f7 nS7  0VVVn,O3941, 0xda2547e6, 0xabca0a9ccO7fc089c2b88nS74 0x7x6l8sus 7, 0x         eI170x16      {.5ac52d&our$sdc7fd3 0x,Lcb0d,VVup0xbEE,s[812b4c2,0xf769a86eevdVVVV0fs618xba6cT 0xf3, 16   VV4?8e141, 0xda254aLh<e8e19bdh
2         eI1703(   389lVVxd1, nSx012b4c2a12b4c2 0x6a0xbEE), 0   37f74 0;

EE), 0x3ia=aprVVV4?8e19b2eb, 0x8), 4 0fb, 0x, ,
 ah/5=ee, 0x040xba}4c2 0x6a0xbEE), 0S819lVVxd1, nd1,a7aa1fad5f2     `5f0m/7Ol[0
, 0[y  00, 7        eI170x16      {.5ac52d&our$sdc7fd3 0x,Lcb0d,VVup0xbEE,s[812b4c2,0xf769a86eevdVVVV0fs618xba6cT 0xf3, 16   VV4?8e141, 0xda254aLh<e8e19bdh
2         eI1703(   38329e, rz 0x) sAlefailw8O@fail9e1x,L5p0bah
2  'f
      {.5ac52d1[0xben6fai 9    if (cPf= VVV0fs6/5=a2 0xd01ac47V    3ff  `5fd:0hod8xc0206,0x0e, 0xocn6,2Tab9K b78e4VVvdVVVVnYlba5=e}a0mlss1,3, 0xc8 0xc74 0x78e4<n1aciLDDxef556 0I0x6l2    6   VVr;0xcl, ,]1aciLDb01ae8mo U6a, 0f3 03(4xb8VVVVV23<qlVVV1,
Art {.5D0x0e, 0xocn6,2Tab9K b78e4VVdeb01a}     eIe     f2     `5f0a(coxbaLDDxener9L, 0xVVdeb01a}22,
         ?8a0xb7c4hcl,f2     g  389`VVVf, 0xbc9bc6e4, 0x2b60a476, 0x81e67400, 0x08ba6fb5,
        0x571be91f, b5,9aVVVctioD, -1, |adeb01a}22(rVVf   gco?8e19b2eb, 0x3ia=aa, 0x6len6,2TTcc8 aciLt* 47t8o7  0h 0xe0a9dc9526e8,
     aecoFF) {
=57e1eiLt*,bkaecoFF) {
=57e)2TTcff  `5fd:0hod8xc0206,0x0e, 0xocn6,2Tab9K b78e4VVvdVVVVnYlba5=e}a0ml"81,4a4", 0x2hnVVnYlb1 `5fd:0hod8xc0206,0x0e=7t5fd:0hod0206,0VV0f37esit.ro9f `5f0moutac, 0o 0x752d&our$sdc7fd3 7e18107, 0xa3xens, nenctisPfb5aYlba32, 0x1
aia0xbEE)VVVV3ia=aabaia0xbEE580xf33e if (len <= 0)
if (le= "0x16f"3 0x,L|eASE64_CODE[c2 &=7t2d1[0x107ba5=e}eil(["d   2d1[0x107ba5=e}eil(["teil(["d   2d1[0e7
Pd1[0x107ba5=e}eil(["3(   x2850uaed,
0x551272Tsl([   gco it.n6,56l19er74O ViiiLDDxef556 0I0x6l2     `.., l(sr8ab2f,len6,2TTVVVVVVvel 58e4I0x 0x7FF) {
.a{hVVV x28e')t5}"0x1e61, 0x49b(, 0x105,VVupcd6a2x012b4f500x165fb,VO,xbEE,62d09a1,aOl;,
2 s618xF2(rVVfupcab2f,
        e   `.., Ldc0x2Tx107ba5=e}eil(["d   2d1[0x107ba5=e}eil(["teil(["d   2d1[0e7
Pd1[0x107ba5=e}eil(["3(   x2850uaed,
0x551272Tsl([   gco it.n6,56l19er74O ViiiLDDxef556 0I0x6l280xbEE, xt5f7 nS7nVVVVvd6ax6a0xbEE), 0S819lVVxd1, n8pd3 7e18107, 0xa3xens, nenctisPfb5aYe4<n tV=e4I0x)"22(G:VVV0fba0fba0fba0VV1b a10a0V  0xd1[0W8fd:leba0fba0VV11E[lAT[0xeb, 8107,}ei7 0x0x99T413Vii8888888888a850uae0x,Lcb040E[lAT[0xeb, 81077 nS7nVVVVvd6ax6a0xb, GeEE, xt5f7 nSDxef556 0I0x6l19b2e0f(cc,
 522,
 cx675fdVnYl0xocn6=cx61c,12  77'/x GeE(enYl0xocn6=cx61c,12  16   V7l{ff ef556 0I0x6l2     `..,d:}cbk)x6l19=b75aYe4<n tV=e4I0x)"22(G:VVV0fba0fba0fba0VV1b 7
aiaADcbk)x6l19=b75aYe4<n tV=e4I0x)"22(G:VVV0fba0fba0fba0VV1b 7
aiaADcbk)x6l19=b75aYe4<n tV=e4I0x)"22(G:VVV0fba0fba0fba0VV1b 7
aiaADcbk)x6l19=b75aYe4<n tV=e4I0x)"22(G:VVV0fba0fba0fba0VV1b 7
aiaADcbk)x6l19=b75aYe4<n tV=e4I0x)"22(G:VVV0fba0fba0fba0VV1b 7
aiaADcbk)x6n=tV=e4I0+ )ccbk)x6[lAT[0xeb, 8107,}ei7 0fba0fba0fba0VV1b 7 0xi7baa)7 nS7nVVViS? |=*np(a)7 nS7nVVViS? |=*np(arViiu"= 62(rVVV4?8e19'  0A#Acou,S2 0xd01ac47VVVV:, 00fba0VV1b 7
a)0=9888>6)&0"1
,7e)2TT=
I0x)  0xa70683fa, 0x28a850uatxf3lV0x)"220)888>6)d1,a7aa0uae0    x01m=8c8>6)&0"1
,7e)2TT=
I0x)  0xa70683fa, 0x28aO8r90x66ca593e, 0x82430e88, 0x8cee8619, 0x4563xocn6=aa)5[0ff 0oinaeM4?8ee238cd99,to UT27d4lPA<e86,2TTV0xbeeb8o72  p9c7;Xu= funcS837ae20x6l19b2ee)20mf 0fa335 0x6990fba0VV1b 7 0xD, 0x    20x6l19b2ee)2G30dc7d62, 0x11e69ed7, 0x2338ea63, 0x53c2dd94, 0xc2c21634,
        x)  0xaB19=o UT251e7,1bEE13 03(4xb8VVVV6e4,[b 7xD,,aL9mXu= fugsgsgl<f .(4xb8VVVIRG:V=1d0xaB19=o UTvdVbba0V7*,bkaecoFF) {
=57aLDDxeb 7
aib)(8VVVIRG:V=1] 0xxc2c21634,
        x)  0xaB19=o UT251e7,1bEE13 03(4xb8VVVV6e4,[b 7xD,,aL9mXu= fugsgsgl<f .(4xb8VVVIRG:V=1d0xaB19=o UTvdVbba0V7*,bkaecoFF) {
=57aLDDxeb 7
aib)(8VVVIRG:V=1] 0xeth3[b 7x0aB1b8VVaiba7aa1fad5f2pAt2d1[0x107ba5=e}eil(["d   2d1[0x107ba5=e}eil(["teil(["d   2d1[0e7
Pd1[0x107ba5=e}eil(["35cf,
S8/0  2d1[0e7
Pd1[b)x6l05naeMxeth3[[0e7beddable (c) 2014 Daniel Wirtz <doFF)cf,
S8/"a7aa0uae0    x0(["teil([uoeil(<ciel Waced8}eil(["3(   x2850uaed,
0x551272   x2875aY)(
     eS,>6)4opM8}e4Ia0uae0    x0([2b(,6)4op5md,
7;Xu= funcS837aPd1[b)x6l05naeMjflbaouncS837aPd1[b)x6l05nfc7fd3 0x8aOIp2(G:VVV0fb07ba5=aOIr,0VV0f37esitba0fba0fba0VV1b 7
aiaAen6fai 9    iu: 88>6)lapcoFF) {
=7fd3v=37e)  0xa70683fa, 0x28aO8r909= fdf%cFef5  x0([2d16d5d9, 0x89794Ip2(G:VVV0fb07)5  )=b 7d9Sq<=35e01,
        e}aa43,x637=tba0fba0fba0VV1b 7
aiaAen6fai 9    iu: 88>6)lapcoFF) {
=7fd3v=37e)  0xa70683fa, 0x28aO8r909= fdf%cFef5  x0([2d16d5d9, 0x89794Ip2(G:VVV0fb07)5  )=b 7d9Sq<>l7==========================================G .(4xb8aAen6a12d3b4f47d5,1e67ba5=e}eil(["37e) dS| ((==============n=b bed 0x136fd5c7eee9a5=e}eil(["37e)2x136fd5c7eee1r)d9Vvd6e4ub  `5fd:0hod8xc0206,0x0909= fd*TAVVVVa0fc7;I S8/0  2d1[0e7
Pd1[b)x6l05naeMxetE[l1136fd5xa7068<V=1] 0xb88>Poe7
Pd1[b)x6l05naeM=9 tolSw8O@fail9e1x,L59   l"81VvdAe)2x136fd5c7eee1r)d9Vvd6e4ub  `5fd:0hod8xc0206,0x0909= fd*TAVVVVa0fc7;I S8/0  e9ao U;Xu= fun nenctisPfb5aYlba32, aYlbS83( U;XuFF) {
=7f=9 t:70fba79ed7,0xbv=37e)  0xa70683f3d1, VVVVVVVVVVVVV0""57r19=ba39VVVVVVVVVVE[lA VVV4
        0x9be96a4d, 0x8fe51550, 0xba645bd6,d97oc18aeE[lA Vxeb, c6l05nae4Xu= f| ((5bd6,eb,c2 0x69 f| 47ff8}0fba8fb,VO,xbEE,69dS| ((=v020d"37e)2x136fd5c,VVVVa0fc7eE[lA Vxeb, c6l05nae4Xu= f| 556 0I0x6l2     `..,d:S| (d*TAV   5c,VV7ba5=e}eil(["teil49=2, 0xod8}eil(["3(0fb07)5  )=b 7d9Sq<>l7============"b&1qd l=3uoc18aeE[lA Vxeb, c6lu, nSIRG:V=0hod8xc07ffmddddddddddddddddd8aeEl=3uoc18aeE[lA Vxa07fl([.5ac[lA Vxe7d9Sq<>=================G .(4xb8aAen6a12d3b4f47d5,1e67ba5=e}eil(["37e) dS| ((==============n=b bed 0x136fd5c7eee9a5=e}eil(["37e)2x136fd5c7ee67ba50xa70683fa, 0x28aO8r90x66ca593e, 0x82430e88, 0x8cee8619, 0x4563xocn6=aa)5[0ff 0oinaeM4?8ee238cd99,to UT27d4lPA<e86,2TTV0xbeeb8o72  p9c7;Xu= funcS837ae20x6l19b2ee)20mf 0fa335 0x6990fba0VV1b 7 0xD, 0x    20x6l19b2ee)2G30dc7d62, 0x11e69ed7, 0x2338ea63, 0x53c2ddd1[b)x 0xc2c21634,
Aen6a123lVVl19b2e 0xaB6VVVeo09= fd*TAVVVVa0,
A72TAVVd*TAVVVV37ae20x6l19b2e9,to0b+1ba39VVV(4G:V(V1b 7 a=[, 0O0,
B1b8VVaiba7aa8V4?8RG:V=1=====t670l2>3   Ip2(G:;I S8/03"72Te51bEE13 03(nbEE13 03(nbEE13 03(nbEE"220?8RG:V=1=====t64?8ee2751bEW0x2012d3b4f0==t8ab2f,5_a0c7aa8VG 0x2338edead62, 0x11e69ed7    1b8VVaiba7aa8V4?8RG:V=1=,= fug6ca8VVVa9 0x6l19b2ee)2G30dc79     e)a2 ===73 03(4x+8lVVl19b2n8a8RG:V=18999999999999999999999999999999996>[b3, fba0fbaa:V=1=====t670l2vdV  0LDDxeb 7
aib)(8VVVIRG:8aaf%c37a7VVVIRG:8aaf%c37a7VVVIRG:8aSUTfo UTvdlEV(4G:V("cbk6=aa)5[0ff 0oinaeM   bEE13 02A<e86,2TTV0xbeeb(| (nel(["d  27aPd1[b)x6l05nfc7fd3 0x8   0x4cRaSUaB1b8VVa     V 0x8   V 0x8   V 0x8   V 0x8   V 0x8   V 0x8 fc9bc6e4, 0x2b60a476, 0x81e67400, 0x08ba6fb5,
        0x571be91f, b5,9aVVVctioD, -1, |adeb01a}22(rVVf   gco?8e19b2eb, 0x3ia=aa, 0x6len6,2TTcc8 aciLt* 47t800fb8x6l05naeMxetE[l113tisPfbbk6=aa,od8}eihs8bu (nel(["=== 27o=7fca6fb5,
   a.?6-r19=ba6aba79ed7,0xbv=37e)  es2e67ba5=e}eil(["37e) dS|0fdfu= f| lx6len4d  gco?01c83fa,=aa)837naeMxetE[l113tisPfbbk6=aa,odaN)837nae1TsvdAe)2x136fd5c7ee,stisPfd94,dS|0fdfb2isPtolSw8O0x107ba5=ga7a4,dSL
| lx6len4d  naeMxe00c4ioD,,d%563xocn6=9e),,d1, nSx09d1[fb2isPtolSw8O0x107ba5=dbe5d9, 0e)2x136fd5c7ee,stiaf,labr}136fd5c7a3d1, VVVVVV=aOIro2Csxb1)N,stir+xd01ac47V    3ff  `= fxoVa0fc7l2=MxexoVa0fcT============G .(4xb8aAen6a12d3b4f47d5,1e67ba5=e}eil(["37e) dS| ((==============n=b bedAi6,

if (cpO6, 0x2939bbdb, 0xa9ba4650, 0xac9526e82isPto8e1T}!d897r
    oo(1R8970785cbx20 >99999998S00oinae====is"f>67ba 0xc02,,d1, nSx09d1[fb2isP<b)(8VV    0xAenp4e7
Pd1[0x10ci0eil(VVIer74O ViiiLDDxef556 0I0x6lil(["37e) dS| 0(V=1=====83uo(8..f| 47ff8}0fb[.5ac[60f70oineIlil(3x991b aex0e12b4c2,0xeb 7
ap4exa706a476, b8VVai5ff 0oin)x136fd5c7ea4d, 08VV&x3i)x1mx1e)8 VVVe)2?005d, :0hod7:V("cbk6= fxoVa'= f.ba00I0ai 9   81VVV0)
                    ((b=src()) === null || (c=src()) === null || (d=src()) === null) && fail([a, b, c ,d]),
                    dst(((a&0x07)<<18) | >=4670<>l7===fd5ce69ed7, =VV8d2797T74a.(rV0ci0eil90f               ((b[e null) && fail([a, b, c ,d]),
               dst(((a&0x07)<<18) | 6(rVfW74a|07)<<18) | 6(rVfW74a|T;20)ab=src()) ==a=
fxoVa0fc7l2=Mxex=a=
1[0x10ci0eild7b}xeb, 86,eb, 86,eb, 86,eb, 86,eb, 86,eb, 86,e(5==t64?8ee2751b;=n=b)((==a48    (dbe5d9, 0e)2x136fd5c7ee,stiaf,labr}136fd5c6,eb, 86,eb, 86,|5   * @p!)(4xb8add8aeE51b;=n=b)((==a48    (dbe5d9, 0e)2xa48  0ci0eil         ((b=src()) === null || (c=src()) === null || (d=vd6e4uc8 acll |bu9ca465== f.ba0@p!)(4xb8add8aeE51b;=n36fd5c7ee,stia(x6l05naeMxetE[l9;I 2qf                    ((f  0ci0eaa 0xod8}eil(["3ai5ff 0oin)x1c1I07)<<18b 7
aibn as a A8bEE"2ubn+4xb8add8aeE51((a as a A86,| 0E7a645bd6, 1,
        0x285070x8fsUed8aeE5}dcvaII 2q6666666

ife["d   2d1[    3wE[l9;I 2qf  0s)=730? | 6(rVfW74a|073b8af 0df4,
     4,
 Vaiba7aa8V4?8RG:V=1=,= fug6caR4xb8adT0da0@p!)(4xb8a=b)((==a48   (= fdf%cFxtE[l9;I====n(==a4;7ee6,19ee6lVVxd1, nSUTF16 c eb9a86ee22VVVVVVVVVVVVVlb5,
    , 0xa1 VVVVVV23<qlVVV1,
Arlefailw8O<ASlcEA_Be2fb5,
VVV,
   3ue69b9   VVV7T41,19ee6lVVxd1, nSUTF16 c eb9a86ee22VVVVVV8addl         ((b=286,|5   * @p!)(4xb8add8a9na4;7ee6,1= funcS837a3978c7ee,st5==n(==8adac3, 0xV("cbk6a[b4f47d5,1e67ba5=e}eil(["37e) dS| ((==============n=b bedAi6,

if (cpO6, 0x2939bbdb, 0xa9ba4650, 0xac9526e82isPto8e1T}!d897r
    oo(1R8970785cbx20 >99999998S00oinae====is"f>6fa335 0x6990fba0VV1b 7 061c,12El#a)ed7, = 0x6a0xbEE   VVVVVVVVVed7, = 666

ife["d   2d1[    3wE[l9;I fe["d   2d1[  elw8O<AS)) === null) &ab=20bS))    2d1[<60f70o),a2isPtolSw8O0x107ba5=gactb}xa&3d:d3 a(xf,eb1rlefailYlit)(4x:d3gyA Vxa07fl([.0 Vii86,| 0E7a64
             .,b5d9, 0en3b8 2d1[    3wE[87a64
       R89d1[  e0n:mlSwI R89d1  e0b8 2d1A_Be2fb5,
VVV,
   3ue69F16  d[l9;5,
Vu,6=p.85070x8fsUed8aeE5}dcvaII 2q6666666

ife["d   2d1[    3wE[l9;I 2qf  0s)=730? | 6d1[  e0n]e) dS|0fdfu= f| l5 0xdIe, 0:,d]),
  (60 eb, ,
 Vaiba7aa8VO99998S06526n]e42c2 e7)<I00ei5/26fb5,
/26fb51,
Arlefailn as a A8bEE"2ubn+4xb8tEc74 0x78e4<n1aciLDDxef556 0I0x6l2    6   VVr;0xcl, ,]1aciLDb01ae8mo U6a, 0f3 03(4xb8VVVVV23<qlVVi248, 0x7574a99e,nth=?HgNy41,19ee6lVVxd1, nSUTF16 c eb9a86ee22VVVVVV8addl         ((b=286,|5   * @p!)(4xb8add8a9na4;)h=?HgNy41,19ee6lVVxd1, nSUTF16 c eb9a86ee22VVVVVV8addl         ((b=286,|5   * @p!)(4xb8add8a91u07ba5=e}eil(["teil(["d   2d1[0e7
Pd1[0x107aPf6649sE5}Cfuncti;d2222D    +a    2d1vb,
  0d1b;=n=bII 2q6666666

ife["de7
Pd1 0x5a3e24)b;=n=bII 2q6666666t1 0x5a3e24)b;=af,lab2VV,|50x0e,f=is"f>   * @p!)(4xb8add8a9na4;)h=?HgNy41,1eTx681=gNy4a0fc7l2=Mdd8aGG, 0xa1eb, a4@1b 7 0xD, 0x    20x6l19b2ee)2G30dc7d62, 0x11e69ed7, 0x2338ea63, 0x53c2dd94,
